### PR TITLE
Removing BinaryOps.SUB to use UnaryOps.NEG and BinaryOps.ADD

### PR DESCRIPTION
**Pull request by sytandas**
Linter woks only remaining: working on it right now. 
Thanks.

### CODE REVIEW

**Comment by geohot** [comment]
BinaryOps.SUB is still being defined

**Comment by sytandas** [comment]
Yes, working on it.

**Comment by sytandas** [comment]
I think it is done, please review if there are some issue still remains.

**Comment by geohot** [comment]
Bounty locked, looking now

File: extra/optimization/helpers.py [hunk 1]
```diff
@@ -30,6 +30,7 @@ def load_worlds(filter_reduce=True, filter_noimage=True, filter_novariable=True)
   if filter_reduce: ast_strs = [x for x in ast_strs if "ReduceOps" in x]
   if filter_noimage: ast_strs = [x for x in ast_strs if "dtypes.image" not in x]
   if filter_novariable: ast_strs = [x for x in ast_strs if "Variable" not in x]
+  ast_strs = [x for x in ast_strs if "BinaryOps.SUB" not in x]
```
Comment (geohot) [inline extra/optimization/helpers.py:33]: Why is this here? You have to regenerate the dataset to not have this.

File: test/test_linearizer.py [hunk 2]
```diff
@@ -219,9 +219,9 @@ def test_reduceops_order(self):
...
     load = MemBuffer(idx=4, dtype=dtypes.float, st=ShapeTracker.from_shape((32,)))
     ast0 = LazyOp(op=ReduceOps.SUM, src=(LazyOp(op=BufferOps.LOAD, src=(), arg=load),), arg=(0,))
-    ast1 = LazyOp(op=ReduceOps.SUM, src=(LazyOp(BinaryOps.SUB, src=(LazyOp(op=BufferOps.LOAD, src=(), arg=load), ast0)),), arg=(0,))
-    ast2 = LazyOp(op=ReduceOps.SUM, src=(LazyOp(BinaryOps.SUB, src=(ast1, LazyOp(op=BufferOps.LOAD, src=(), arg=load))),), arg=(0,))
-    ast3 = LazyOp(op=ReduceOps.SUM, src=(LazyOp(BinaryOps.ADD, src=(LazyOp(op=BinaryOps.SUB, src=(LazyOp(op=BufferOps.LOAD, src=(), arg=load), ast2)), LazyOp(op=BinaryOps.SUB, src=(LazyOp(op=BufferOps.LOAD, src=(), arg=load), ast0)),)),), arg=(0,)) # noqa E501
+    ast1 = LazyOp(op=ReduceOps.SUM, src=(LazyOp(BinaryOps.ADD, src=(LazyOp(op=BufferOps.LOAD, src=(), arg=load), LazyOp(op=UnaryOps.NEG, src=(ast0,), arg=None))),), arg=(0,)) # noqa: E501
+    ast2 = LazyOp(op=ReduceOps.SUM,  src=(LazyOp(BinaryOps.ADD, src=(ast1, LazyOp(op=UnaryOps.NEG, src=(LazyOp(op=BufferOps.LOAD, src=(), arg=load),), arg=None))),), arg=(0,)) # noqa: E501
```
Comment (geohot) [inline test/test_linearizer.py:223]: Do these need noqa? Why is there a random space?

The space itself doesn't matter, but it shows me you haven't carefully read this diff.

File: test/test_linearizer_failures.py [hunk 3]
```diff
@@ -121,7 +118,7 @@ def test_failure_14(self):
...
...MemBuffer(idx=4, dtype=dtypes.float, st=ShapeTracker( views=(View(shape=(1, 1, 112, 14, 14, 1, 1, 1), strides=(0, 0, 1, 0, 0, 0, 0, 0), offset=0, mask=None, contiguous=False),))))), arg=None), LazyOp(op=UnaryOps.SQRT, src=(LazyOp(op=UnaryOps.RECIP, src=(LazyOp(op=BinaryOps.ADD, src=( LazyOp(op=BufferOps.LOAD, src=(), arg=MemBuffer(idx=5, dtype=dtypes.float, st=ShapeTracker(views=( View(shape=(1, 1, 112, 14, 14, 1, 1, 1), strides=(0, 0, 1, 0, 0, 0, 0, 0), offset=0, mask=None, contiguous=False),)))), LazyOp(op=BufferOps.CONST, src=(), arg=ConstBuffer(val=1e-05, dtype=dtypes.float, st=ShapeTracker(views=(View( shape=( 1, 1, 112, 14, 14, 1, 1, 1), strides=( 0, 0, 0, 0, 0, 0, 0, 0), offset=0, mask=None, contiguous=False),))))), arg=None),), arg=None),), arg=None)), arg=None), LazyOp(op=BufferOps.LOAD, src=(), arg=MemBuffer(idx=6, dtype=dtypes.float, st=ShapeTracker( views=(View(shape=(1, 1, 112, 14, 14, 1, 1, 1), strides=(0, 0, 1, 0, 0, 0, 0, 0), offset=0, mask=None, contiguous=False),))))), arg=None),), arg=MemBuffer(idx=0, dtype=dtypes.float, st=ShapeTracker(views=( View(shape=(1, 1, 112, 14, 14, 1, 1, 1), strides=(0, 0, 196, 14, 1, 0, 0, 0), offset=0, mask=None, contiguous=True),))))
```
Comment (geohot) [inline test/test_linearizer_failures.py:121]: Again, why is spacing changing?

Comment (sytandas) [reply test/test_linearizer_failures.py:121]: I will correct it asap, will be back within hrs. Thanks.

File: test/test_linearizer.py (hunk 2 above)
Comment (sytandas) [reply test/test_linearizer.py:223]: yes correct it asap.

**Comment by geohot** [comment]
Locked to #4977 instead. It already has the dataset regenerated and has many fewer cleaner changes.

**Comment by geohot** [comment]
One example of why #4977 is better.
Screenshot 2024-06-17 at 6 17 41 AM
Screenshot 2024-06-17 at 6 17 44 AM

See how it keeps the tests.

**Comment by geohot** [comment]
Also note the differences in number of commits.
Screenshot 2024-06-17 at 6 20 04 AM
Screenshot 2024-06-17 at 6 20 14 AM

Again, it doesn't really mean anything, but it's a good proxy for quality.

**Comment by sytandas** [comment]
Yes I want to learn and collaborate, surely I do make some mistake. But I will learn from you guys to get better.  
I will learn from it, thank you George.

**Comment by sytandas** [comment]
Yes I want to learn and collaborate, surely I do make some mistake. But I will learn from you guys to get better.
No doubt he done it better than mine.
I will learn from it will be get better, thank you.

**Comment by geohot** [comment]
Bounty awarded to #4977
